sdk/src/types/block/address/multi.rs

```diff
@@ -0,0 +1,226 @@
+use iterator_sorted::is_unique_sorted;
+use packable::{
+    bounded::BoundedU8,
+    error::{UnpackError, UnpackErrorExt},
+    packer::Packer,
+    prefix::BoxedSlicePrefix,
+    unpacker::Unpacker,
+    Packable, PackableExt,
+};
+
+use crate::types::block::{address::Address, Error};
+
+pub(crate) type WeightedAddressCount =
+    BoundedU8<{ *MultiAddress::ADDRESSES_COUNT.start() }, { *MultiAddress::ADDRESSES_COUNT.end() }>;
+
+/// An address with an assigned weight.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, From, AsRef, Packable)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct WeightedAddress {
+    /// The unlocked address.
+    #[packable(verify_with = verify_address)]
+    address: Address,
+    /// The weight of the unlocked address.
+    #[packable(verify_with = verify_weight)]
+    weight: u8,
+}
+
+impl WeightedAddress {
+    /// Creates a new [`WeightedAddress`].
+    pub fn new(address: Address, weight: u8) -> Result<WeightedAddress, Error> {
+        verify_address::<true>(&address, &())?;
+        verify_weight::<true>(&weight, &())?;
+
+        Ok(Self { address, weight })
+    }
+
+    /// Returns the address of the [`WeightedAddress`].
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Returns the weight of the [`WeightedAddress`].
+    pub fn weight(&self) -> u8 {
+        self.weight
+    }
+}
+
+fn verify_address<const VERIFY: bool>(address: &Address, _visitor: &()) -> Result<(), Error> {
+    if VERIFY {
+        if !matches!(
+            address,
+            Address::Ed25519(_) | Address::Account(_) | Address::Nft(_) | Address::Anchor(_)
+        ) {
+            return Err(Error::InvalidAddressKind(address.kind()));
+        }
+    }
+    Ok(())
+}
+
+fn verify_weight<const VERIFY: bool>(weight: &u8, _visitor: &()) -> Result<(), Error> {
+    if VERIFY && *weight == 0 {
+        return Err(Error::InvalidAddressWeight(*weight));
+    } else {
+        Ok(())
+    }
+}
+
+/// An address that consists of addresses with weights and a threshold value.
+/// The Multi Address can be unlocked if the cumulative weight of all unlocked addresses is equal to or exceeds the
+/// threshold.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MultiAddress {
+    /// The weighted unlocked addresses.
+    addresses: BoxedSlicePrefix<WeightedAddress, WeightedAddressCount>,
+    /// The threshold that needs to be reached by the unlocked addresses in order to unlock the multi address.
+    threshold: u16,
+}
+
+impl MultiAddress {
+    /// The [`Address`](crate::types::block::address::Address) kind of an [`MultiAddress`].
+    pub const KIND: u8 = 40;
+    /// The allowed range of inner [`Address`]es.
+    pub const ADDRESSES_COUNT: RangeInclusive<u8> = 1..=10;
+
+    /// Creates a new [`MultiAddress`].
+    #[inline(always)]
+    pub fn new(addresses: Vec<WeightedAddress>, threshold: u16) -> Result<Self, Error> {
```

Comment:
impl Iter instead? then collect to boxed slice?

Reply:
Iterator or IntoIterator?

Reply:
probably IntoIterator